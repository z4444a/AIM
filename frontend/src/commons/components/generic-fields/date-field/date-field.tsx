import React, { ReactNode } from 'react';
import {
  DateFieldConfigModel,
  DateFieldHandlers,
  DateFieldValue,
} from '../models/field-config.model';
import Tooltip from '@material-ui/core/Tooltip';
import moment from 'moment';
import { MaterialUiPickersDate } from 'material-ui-pickers/typings/date';
import AimDatePicker from '../../date-picker/aim-date-picker';
import AimMultipleDatePicker from '../../aim-multiple-date-picker/index';

export interface Props {
  fieldConfig: DateFieldConfigModel;
  handlers: DateFieldHandlers;
  value?: DateFieldValue;
  values?: DateFieldValue[];
  today?: boolean;
  needTooltip?: boolean;
}

export type InternalProps = Props;

export class DateField extends React.PureComponent<InternalProps> {
  public render(): ReactNode {
    const { fieldConfig, value, today, needTooltip, values } = this.props;
    const {
      className,
      style,
      required,
      labelConfig,
      error,
      FormHelperTextProps,
      disabled,
      multipleMax,
    } = fieldConfig;
    let minDate, maxDate;
    if (fieldConfig.constraints) {
      const { minDateValue, maxDateValue } = fieldConfig.constraints;
      minDate = minDateValue;
      maxDate = maxDateValue;
    }
    const multiple = multipleMax && multipleMax > 1;
    const pickerProps = {
      className: className,
      style: style,
      label: labelConfig && labelConfig.visible ? labelConfig.value : undefined,
      required: required,
      error: !!error,
      helperText: error ? error.value : undefined,
      emptyLabel: today ? '"сегодня"' : '',
      minDate: minDate,
      maxDate: maxDate,
      FormHelperTextProps: FormHelperTextProps,
      disabled: today ? true : disabled,
    };
    return (
      <Tooltip
        title={
          needTooltip
            ? fieldConfig.constraints
              ? this.getTooltip(
                  fieldConfig.constraints.minDateValue,
                  fieldConfig.constraints.maxDateValue
                )
              : this.getTooltip()
            : ''
        }
      >
        {multiple ? (
          <AimMultipleDatePicker
            values={values ? values : []}
            valueLimit={multipleMax}
            onValuesChange={this.handleValuesChange}
            value={undefined}
            onChange={() => {}}
            {...pickerProps}
          />
        ) : (
          <AimDatePicker value={value || null} onChange={this.handleValueChange} {...pickerProps} />
        )}
      </Tooltip>
    );
  }

  private handleValuesChange = (values: Date[]) => {
    const { handlers, fieldConfig } = this.props;
    if (!handlers || typeof handlers.onValuesChange !== 'function') {
      return;
    }
    handlers.onValuesChange(values, fieldConfig.key);
  };

  private handleValueChange = (date: MaterialUiPickersDate) => {
    const { handlers, fieldConfig } = this.props;
    if (!handlers || typeof handlers.onValueChange !== 'function') {
      return;
    }
    handlers.onValueChange(date.toDate(), fieldConfig.key);
  };

  private getTooltip = (
    min: Date | undefined = undefined,
    max: Date | undefined = undefined
  ): string => {
    return (
      'Выберите дату' +
      (!min && !max
        ? ''
        : min && max
        ? ' из диапазона ' +
          moment(min).format('YYYY-MM-DD') +
          ' - ' +
          moment(max).format('YYYY-MM-DD')
        : min
        ? '. Дата должна быть не раньше ' + moment(min).format('YYYY-MM-DD')
        : '. Дата должна быть не позже ' + moment(max).format('YYYY-MM-DD'))
    );
  };
}
