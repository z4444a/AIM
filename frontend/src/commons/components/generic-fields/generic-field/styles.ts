import { Theme, WithStyles } from '@material-ui/core';
import createStyles from '@material-ui/core/styles/createStyles';

export type Style = WithStyles<'container'>;

export const styles = (theme: Theme) =>
  createStyles({
    container: {
      width: '100%',
      paddingTop: theme.spacing.unit,
      '& > *': {
        width: '100%',
      },
    },
  });
