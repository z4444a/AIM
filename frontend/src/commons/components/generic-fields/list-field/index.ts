import { ListField } from './list-field';

export default ListField;
