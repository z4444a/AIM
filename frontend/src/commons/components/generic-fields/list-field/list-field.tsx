import React, { ReactNode } from 'react';
import TextField from '@material-ui/core/TextField';
import MenuItem from '@material-ui/core/MenuItem';
import {
  ListFieldConfigModel,
  ListFieldHandlers,
  ListFieldValue,
} from '../models/field-config.model';
import AimMultipleSelect from '../../aim-multiple-select/index';
import NamedModel from '../../../../model/base/named-model';
import { ItemSelection } from '../../aim-multiple-select/aim-multiple-select';
import _ from 'underscore';

export interface Props {
  fieldConfig: ListFieldConfigModel;
  handlers: ListFieldHandlers;
  value?: ListFieldValue;
  values?: ListFieldValue[];
}

export type InternalProps = Props;

export class ListField extends React.PureComponent<InternalProps, ItemSelection> {
  private readonly emptyValue = '';

  constructor(props: InternalProps) {
    super(props);
    this.state = ListField.getSelection(
      props.fieldConfig.optionList,
      props.values ? props.values.map(item => item.toString()) : []
    );
  }

  componentDidUpdate(prevProps: Readonly<InternalProps>, prevState: Readonly<ItemSelection>): void {
    const { values, fieldConfig } = this.props;
    if (
      !_.isEqual(values, prevProps.values) ||
      !_.isEqual(fieldConfig.optionList, prevProps.fieldConfig.optionList)
    ) {
      this.setState(
        ListField.getSelection(
          fieldConfig.optionList,
          values ? values.map(item => item.toString()) : []
        )
      );
    }
  }

  public render(): ReactNode {
    const { fieldConfig, value } = this.props;
    const {
      className,
      style,
      required,
      key,
      optionList,
      labelConfig,
      error,
      multipleMax,
    } = fieldConfig;
    const multiple = multipleMax && multipleMax > 1;
    const { available, selected } = this.state;
    return multiple ? (
      <AimMultipleSelect
        disabled={selected.length === multipleMax || available.length === 0}
        available={available}
        selected={selected}
        selectItems={this.selectItems}
        label={labelConfig && labelConfig.visible ? labelConfig.value : ''}
        error={!!error}
      />
    ) : (
      <TextField
        select
        style={style}
        className={className}
        required={required}
        error={!!error}
        helperText={error ? error.value : undefined}
        value={value ? value : this.emptyValue}
        onChange={this.handleValueChange}
        label={labelConfig && labelConfig.visible ? labelConfig.value : undefined}
      >
        {Object.keys(optionList).map(itemKey => (
          <MenuItem key={`${key}_${itemKey}`} value={itemKey}>
            {optionList[itemKey]}
          </MenuItem>
        ))}
      </TextField>
    );
  }

  private static getSelection = (
    options: { [key: string]: string },
    values: ListFieldValue[]
  ): ItemSelection => {
    const indexList = Object.keys(options);
    return {
      available: indexList
        .filter(item => values.indexOf(item) < 0)
        .map(item => ListField.toNamed(options, item)),
      selected: indexList
        .filter(item => values.indexOf(item) >= 0)
        .map(item => ListField.toNamed(options, item)),
    };
  };

  private static toNamed = (options: { [key: string]: string }, value: string): NamedModel => {
    return {
      id: parseInt(value),
      name: options[value],
    };
  };

  private selectItems = (selection: ItemSelection) => {
    this.setState(selection);
    const { handlers, fieldConfig } = this.props;
    if (!handlers || typeof handlers.onValuesChange !== 'function') {
      return;
    }
    handlers.onValuesChange(selection.selected.map(item => item.id.toString()), fieldConfig.key);
  };

  private handleValueChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { handlers, fieldConfig } = this.props;
    if (!handlers || typeof handlers.onValueChange !== 'function') {
      return;
    }
    handlers.onValueChange(event.target.value, fieldConfig.key);
  };
}
