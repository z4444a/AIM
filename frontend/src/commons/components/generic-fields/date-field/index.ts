import { DateField } from './date-field';

export default DateField;
