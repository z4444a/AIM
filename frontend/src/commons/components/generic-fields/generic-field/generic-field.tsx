import React, { ReactNode } from 'react';
import {
  BaseFieldConfigI,
  DateFieldConfigModel,
  DateFieldValue,
  FieldType,
  GenericFieldHandlers,
  GenericFieldValue,
  ListFieldConfigModel,
  ListFieldValue,
  NumberFieldConfigModel,
  NumberFieldValue,
  RealNumberFieldConfigModel,
  TextFieldConfigModel,
  TextFieldValue,
} from '../models/field-config.model';
import TextField from '../text-field/index';
import NumberField from '../number-field/index';
import DateField from '../date-field/index';
import ListField from '../list-field/index';
import { Style } from './styles';
import RealNumberField from '../real-number-field/index';

export interface Props {
  fieldConfig: BaseFieldConfigI;
  handlers: GenericFieldHandlers;
  value: GenericFieldValue[];
  needTooltip?: boolean;
  toolTipTitle?: string;
}

export type InternalProps = Props & Style;

export class GenericField extends React.PureComponent<InternalProps> {
  private readonly handlers: GenericFieldHandlers;

  public constructor(props: InternalProps) {
    super(props);
    this.handlers = this.getHandlers();
  }

  public componentDidMount(): void {
    const { fieldConfig, handlers } = this.props;
    const list = (fieldConfig as ListFieldConfigModel).optionList;
    const typeId = (fieldConfig as ListFieldConfigModel).typeId;
    if (
      handlers.onFieldRender &&
      typeof handlers.onFieldRender === 'function' &&
      !!typeId &&
      !!list
    ) {
      handlers.onFieldRender(typeId);
    }
  }

  public render(): ReactNode {
    const { fieldConfig, value, classes, needTooltip, toolTipTitle } = this.props;
    const { fieldType } = fieldConfig;

    let field: ReactNode;
    switch (fieldType) {
      case FieldType.Text:
        field = (
          <TextField
            fieldConfig={fieldConfig as TextFieldConfigModel}
            value={value[0] as TextFieldValue}
            values={value as TextFieldValue[]}
            handlers={this.handlers}
            needTooltip={needTooltip}
          />
        );
        break;
      case FieldType.Number:
        field = (
          <NumberField
            fieldConfig={fieldConfig as NumberFieldConfigModel}
            value={value[0] as NumberFieldValue}
            values={value as NumberFieldValue[]}
            handlers={this.handlers}
            needTooltip={needTooltip}
            toolTipTitle={toolTipTitle}
          />
        );
        break;
      case FieldType.Date:
        field = (
          <DateField
            fieldConfig={fieldConfig as DateFieldConfigModel}
            value={value[0] as DateFieldValue}
            values={value as DateFieldValue[]}
            handlers={this.handlers}
            needTooltip={needTooltip}
          />
        );
        break;
      case FieldType.List:
        field = (
          <ListField
            fieldConfig={fieldConfig as ListFieldConfigModel}
            value={value[0] as ListFieldValue}
            values={value as ListFieldValue[]}
            handlers={this.handlers}
          />
        );
        break;
      case FieldType.Real:
        field = (
          <RealNumberField
            fieldConfig={fieldConfig as RealNumberFieldConfigModel}
            value={value[0] as NumberFieldValue}
            values={value as NumberFieldValue[]}
            handlers={this.handlers}
            needTooltip={needTooltip}
          />
        );
        break;
    }
    return <div className={classes.container}>{field}</div>;
  }

  private getHandlers(): GenericFieldHandlers {
    const handlers: GenericFieldHandlers = {
      onValueChange: (value, key) => {
        const { handlers } = this.props;
        if (handlers && typeof handlers.onValueChange === 'function') {
          handlers.onValueChange(value, key);
        }
      },
      onValuesChange: (value, key) => {
        const { handlers } = this.props;
        if (handlers && handlers.onValuesChange && typeof handlers.onValueChange === 'function') {
          handlers.onValuesChange(value, key);
        }
      },
    };

    return handlers;
  }
}
