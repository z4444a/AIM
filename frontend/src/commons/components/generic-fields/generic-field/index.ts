import { GenericField } from './generic-field';
import withStyles from '@material-ui/core/styles/withStyles';
import { styles } from './styles';

export default withStyles(styles)(GenericField);
