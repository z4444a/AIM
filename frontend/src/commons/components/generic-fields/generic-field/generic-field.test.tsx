import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import {
  BaseFieldConfigI,
  DateFieldConfigModel,
  FieldType,
  GenericFieldHandlers,
  ListFieldConfigModel,
  NumberFieldConfigModel,
  RealNumberFieldConfigModel,
  TextFieldConfigModel,
} from '../models/field-config.model';
import { GenericField, InternalProps } from './generic-field';
import TextField from '../text-field/index';
import NumberField from '../number-field/index';
import { DateField } from '../date-field/date-field';
import { ListField } from '../list-field/list-field';
import RealNumberField from '../real-number-field/index';

describe('GenericField', () => {
  function getComponentProps(fieldType: FieldType): InternalProps {
    const handlers: GenericFieldHandlers = {
      onValueChange: () => {},
    };
    let fieldConfig: BaseFieldConfigI = new TextFieldConfigModel('key');
    switch (fieldType) {
      case FieldType.Date:
        fieldConfig = new DateFieldConfigModel('key');
        break;
      case FieldType.List:
        fieldConfig = new ListFieldConfigModel('key');
        break;
      case FieldType.Number:
        fieldConfig = new NumberFieldConfigModel('key');
        break;
      case FieldType.Text:
        fieldConfig = new TextFieldConfigModel('key');
        break;
      case FieldType.Real:
        fieldConfig = new RealNumberFieldConfigModel('key');
    }
    const value = [undefined];
    const props: InternalProps = {
      classes: {
        container: 'container',
      },
      value,
      handlers,
      fieldConfig,
    };

    return props;
  }

  function getShallowWrapper(fieldType: FieldType) {
    const wrapper = shallow(<GenericField {...getComponentProps(fieldType)} />);
    return wrapper;
  }

  it('should render correctly', () => {
    const wrapper = getShallowWrapper(FieldType.Text);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should render TextField', () => {
    const wrapper = getShallowWrapper(FieldType.Text);
    expect(wrapper.find(TextField).exists()).toBe(true);
  });

  it('should render NumberField', () => {
    const wrapper = getShallowWrapper(FieldType.Number);
    expect(wrapper.find(NumberField).exists()).toBe(true);
  });

  it('should render DateField', () => {
    const wrapper = getShallowWrapper(FieldType.Date);
    expect(wrapper.find(DateField).exists()).toBe(true);
  });

  it('should render ListField', () => {
    const wrapper = getShallowWrapper(FieldType.List);
    expect(wrapper.find(ListField).exists()).toBe(true);
  });

  it('should render RealNumberField', () => {
    const wrapper = getShallowWrapper(FieldType.Real);
    expect(wrapper.find(RealNumberField).exists()).toBe(true);
  });
});
