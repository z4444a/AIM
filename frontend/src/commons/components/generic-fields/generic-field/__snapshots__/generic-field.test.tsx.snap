// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericField should render correctly 1`] = `
<div
  className="container"
>
  <withI18nextTranslation(TextField)
    fieldConfig={
      TextFieldConfigModel {
        "FormHelperTextProps": undefined,
        "InputProps": undefined,
        "className": undefined,
        "constraints": undefined,
        "disabled": undefined,
        "error": undefined,
        "fieldType": "Text",
        "fullWidth": undefined,
        "inputRef": undefined,
        "key": "key",
        "labelConfig": undefined,
        "multipleMax": 1,
        "required": false,
        "rootRef": undefined,
        "showHelperOnError": true,
        "style": undefined,
      }
    }
    handlers={
      Object {
        "onValueChange": [Function],
        "onValuesChange": [Function],
      }
    }
    values={
      Array [
        undefined,
      ]
    }
  />
</div>
`;
